lib/guard_against_physical_delete/support_counter_cache/associations/builder/belongs_to.rb

```diff
@@ -27,29 +24,56 @@ def belongs_to_counter_cache_after_logical_delete(reflection)
                   def belongs_to_counter_cache_after_revive(reflection)
                     cache_column = reflection.counter_cache_column
                     delete_column = self.logical_delete_column
-                    return unless send("#{delete_column}_changed?")
-                    return if send("#{delete_column}_was").nil?
+                    return unless saved_change_to_logical_delete_column?(delete_column)
+                    return if logical_delete_column_before_last_save(delete_column).nil?
                     return unless send(delete_column).nil?
                     record = send(reflection.name)
                     record.class.increment_counter(cache_column, record.id) unless record.nil?
                   end
+
+                  if ActiveRecord.version < Gem::Version.new('5.1.0')
+                    def saved_change_to_logical_delete_column?(delete_column)
+                      __send__("#{delete_column}_changed?")
+                    end
+
+                    def logical_delete_column_before_last_save(delete_column)
+                      __send__("#{delete_column}_was")
+                    end
+                  else
+                    def saved_change_to_logical_delete_column?(delete_column)
+                      __send__("saved_change_to_#{delete_column}")
```

Comment:
`attribute_changed?` seems to be renamed to `will_save_change_to_attribute?`, is this intended change?
https://github.com/rails/rails/pull/25337/files#diff-60c25306d36b9a651f87d1c4bb381b33be05620f6fc17ec2ddb77b2ba0011681R224
I want some explanation about this 🤔

Reply:
`attribute_changed?` has been split into two methods, `will_save_change_to_attribute?` and `saved_change_to_attribute?`. See how the behavior of these methods in before/after callbacks changes on different ActiveRecord versions.

```ruby
# attr.rb

require 'rspec'
require 'spec_helper'

puts "# activerecord #{ActiveRecord.version}"
puts

def print_attr_states(at, r)
  ar_5_0 = ActiveRecord.version.to_s < '5.1.0'
  puts <<~END
    # #{at}
    attribute_changed?             : #{r.attribute_changed?(:deleted_at).inspect}
    will_save_change_to_attribute? : #{ar_5_0 ? '(undef)' : r.will_save_change_to_attribute?(:deleted_at).inspect}
    saved_change_to_attribute?     : #{ar_5_0 ? '(undef)' : r.saved_change_to_attribute?(:deleted_at).inspect}

  END
end

Logical.before_update { |r| print_attr_states('before_update', r) }
Logical.after_update { |r| print_attr_states('after_update', r) }

r = Logical.create
r.soft_delete
```

```
% bundle exec ruby -I spec attr.rb
# activerecord 5.0.7.2

# before_update
attribute_changed?             : true
will_save_change_to_attribute? : (undef)
saved_change_to_attribute?     : (undef)

# after_update
attribute_changed?             : true
will_save_change_to_attribute? : (undef)
saved_change_to_attribute?     : (undef)
```

```
% bundle exec ruby -I spec attr.rb
# activerecord 5.1.7

# before_update
attribute_changed?             : true
will_save_change_to_attribute? : true
saved_change_to_attribute?     : false

DEPRECATION WARNING: The behavior of `attribute_changed?` inside of after callbacks will be changing in the next version of Rails. The new return value will reflect the behavior of calling the method after `save` returned (e.g. the opposite of what it returns now). To maintain the current behavior, use `saved_change_to_attribute?` instead. (called from print_attr_states at attr.rb:11)
# after_update
attribute_changed?             : true
will_save_change_to_attribute? : nil
saved_change_to_attribute?     : true
```

```
% bundle exec ruby -I spec attr.rb
# activerecord 5.2.6

# before_update
attribute_changed?             : true
will_save_change_to_attribute? : true
saved_change_to_attribute?     : false

# after_update
attribute_changed?             : false
will_save_change_to_attribute? : nil
saved_change_to_attribute?     : true
```

Regarding guard_against_physical_delete, we need to know if the logical delete attribute have been changed in the after_update callback. There, `will_save_change_to_attribute?` is designed to return nil from the beginning, and `attribute_changed?` always returns false regardless of changes starting from AR 5.2. So `saved_change_to_attribute?` is the only option.